llvm/test/Transforms/InstCombine/commutative-operation-over-selects.ll

```diff
@@ -0,0 +1,1117 @@
+; CHECK-NEXT:    ret i16 [[RET]]
+;
+  %s0 = select i1 %c, i16 %a, i16 %b
+  %s1 = select i1 %c1, i16 %b, i16 %a
+  %ret = call i16 @llvm.umul.fix.i16(i16 %s1, i16 %s0, i32 1)
+  ret i16 %ret
+}
+
+define i16 @fold_select_smul_fix_sat_neg(i1 %c, i1 %c1, i16 %a, i16 %b, i32 %y) {
+; CHECK-LABEL: define i16 @fold_select_smul_fix_sat_neg(
+; CHECK-SAME: i1 [[C:%.*]], i1 [[C1:%.*]], i16 [[A:%.*]], i16 [[B:%.*]], i32 [[Y:%.*]]) {
+; CHECK-NEXT:    [[S0:%.*]] = select i1 [[C]], i16 [[A]], i16 [[B]]
+; CHECK-NEXT:    [[S1:%.*]] = select i1 [[C1]], i16 [[B]], i16 [[A]]
+; CHECK-NEXT:    [[RET:%.*]] = call i16 @llvm.smul.fix.sat.i16(i16 [[S1]], i16 [[S0]], i32 1)
+; CHECK-NEXT:    ret i16 [[RET]]
+;
+  %s0 = select i1 %c, i16 %a, i16 %b
+  %s1 = select i1 %c1, i16 %b, i16 %a
+  %ret = call i16 @llvm.smul.fix.sat.i16(i16 %s1, i16 %s0, i32 1)
+  ret i16 %ret
+}
+
+define i16 @fold_select_umul_fix_sat_neg(i1 %c, i1 %c1, i16 %a, i16 %b, i32 %y) {
+; CHECK-LABEL: define i16 @fold_select_umul_fix_sat_neg(
+; CHECK-SAME: i1 [[C:%.*]], i1 [[C1:%.*]], i16 [[A:%.*]], i16 [[B:%.*]], i32 [[Y:%.*]]) {
+; CHECK-NEXT:    [[S0:%.*]] = select i1 [[C]], i16 [[A]], i16 [[B]]
+; CHECK-NEXT:    [[S1:%.*]] = select i1 [[C1]], i16 [[B]], i16 [[A]]
+; CHECK-NEXT:    [[RET:%.*]] = call i16 @llvm.umul.fix.sat.i16(i16 [[S1]], i16 [[S0]], i32 1)
+; CHECK-NEXT:    ret i16 [[RET]]
+;
+  %s0 = select i1 %c, i16 %a, i16 %b
+  %s1 = select i1 %c1, i16 %b, i16 %a
+  %ret = call i16 @llvm.umul.fix.sat.i16(i16 %s1, i16 %s0, i32 1)
+  ret i16 %ret
+}
+
+define float @fold_select_fma_neg(i1 %c, i1 %c1, float %a, float %b, float %y) {
+; CHECK-LABEL: define float @fold_select_fma_neg(
+; CHECK-SAME: i1 [[C:%.*]], i1 [[C1:%.*]], float [[A:%.*]], float [[B:%.*]], float [[Y:%.*]]) {
+; CHECK-NEXT:    [[S0:%.*]] = select i1 [[C]], float [[A]], float [[B]]
+; CHECK-NEXT:    [[S1:%.*]] = select i1 [[C1]], float [[B]], float [[A]]
+; CHECK-NEXT:    [[RET:%.*]] = call float @llvm.fma.f32(float [[S1]], float [[S0]], float [[Y]])
+; CHECK-NEXT:    ret float [[RET]]
+;
+  %s0 = select i1 %c, float %a, float %b
+  %s1 = select i1 %c1, float %b, float %a
+  %ret = call float @llvm.fma.f32(float %s1, float %s0, float %y)
+  ret float %ret
+}
+
+define float @fold_select_fmuladd_neg(i1 %c, i1 %c1, float %a, float %b, float %y) {
+; CHECK-LABEL: define float @fold_select_fmuladd_neg(
+; CHECK-SAME: i1 [[C:%.*]], i1 [[C1:%.*]], float [[A:%.*]], float [[B:%.*]], float [[Y:%.*]]) {
+; CHECK-NEXT:    [[S0:%.*]] = select i1 [[C]], float [[A]], float [[B]]
+; CHECK-NEXT:    [[S1:%.*]] = select i1 [[C1]], float [[B]], float [[A]]
+; CHECK-NEXT:    [[RET:%.*]] = call float @llvm.fmuladd.f32(float [[S1]], float [[S0]], float [[Y]])
+; CHECK-NEXT:    ret float [[RET]]
+;
+  %s0 = select i1 %c, float %a, float %b
+  %s1 = select i1 %c1, float %b, float %a
+  %ret = call float @llvm.fmuladd.f32(float %s1, float %s0, float %y)
+  ret float %ret
+}
```

Comment:
Add a negative test with a non-commutative operator, e.g. `sub`. Unless I missed it somewhere.

Reply:
Done